File: objective-c/Chat/iOS/Classes/LoginController.m
// **********************************************************************
//
// Copyright (c) 2003-2016 ZeroC, Inc. All rights reserved.
//
// **********************************************************************

#import <LoginController.h>
#import <ChatController.h>

#import <objc/Ice.h>
#import <objc/Glacier2.h>

@interface LoginController()

@property (nonatomic) UITextField* currentField;
@property (nonatomic) NSString* oldFieldValue;
@property (nonatomic) WaitAlert* waitAlert;
@property (nonatomic) id<ICECommunicator> communicator;

@end

@implementation LoginController

@synthesize currentField;
@synthesize oldFieldValue;
@synthesize waitAlert;
@synthesize communicator;

static NSString* usernameKey = @"usernameKey";
static NSString* passwordKey = @"passwordKey";
static NSString* sslKey = @"sslKey";

+(void)initialize
{
    NSDictionary* appDefaults = [NSDictionary dictionaryWithObjectsAndKeys:@"", usernameKey,
                                 @"", passwordKey,
                                 @"YES", sslKey,
                                 nil];
    
    [[NSUserDefaults standardUserDefaults] registerDefaults:appDefaults];
}

- (void)viewDidLoad
{
    // Register IceSSL plugin and load it on communicator initialization.
    ICEregisterIceSSL(YES);

    NSUserDefaults* defaults = [NSUserDefaults standardUserDefaults];

    // Set the default values, and show the clear button in the text field.
    usernameField.text =  [defaults stringForKey:usernameKey];
    usernameField.clearButtonMode = UITextFieldViewModeWhileEditing;
    passwordField.text = [defaults stringForKey:passwordKey];
    passwordField.clearButtonMode = UITextFieldViewModeWhileEditing;
    
    chatController = [[ChatController alloc] initWithNibName:@"ChatView" bundle:nil];

    [[NSNotificationCenter defaultCenter] addObserver:self
                                             selector:@selector(applicationDidEnterBackground) 
                                                 name:UIApplicationDidEnterBackgroundNotification
                                               object:nil];
}

-(void) connecting:(BOOL)v
{
    // Show the wait alert.
    statusLabel.text = @"Connecting...";

    statusLabel.hidden = !v;
    if(v)
    {
        [statusActivity startAnimating];
    }
    else
    {
        [statusActivity stopAnimating];
    }
    loginButton.enabled = !v;
    usernameField.enabled = !v;
    passwordField.enabled = !v;
}

- (void)applicationDidEnterBackground
{
    // Start the long-running task and return immediately.
    dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{
        [communicator destroy];
    });
}

- (void)viewWillAppear:(BOOL)animated
{
    loginButton.enabled = usernameField.text.length > 0;
    [loginButton setAlpha:loginButton.enabled ? 1.0 : 0.5];
	[super viewWillAppear:animated];
}

- (BOOL)shouldAutorotateToInterfaceOrientation:(UIInterfaceOrientation)interfaceOrientation
{
	// Return YES for supported orientations
	return (interfaceOrientation == UIInterfaceOrientationPortrait);
}

- (void)didReceiveMemoryWarning
{
	[super didReceiveMemoryWarning]; // Releases the view if it doesn't have a superview
	// Release anything that's not essential, such as cached data
}

#pragma mark UITextFieldDelegate

-(BOOL)textFieldShouldBeginEditing:(UITextField*)field
{
    self.currentField = field;
    self.oldFieldValue = field.text;
    return YES;
}

-(BOOL)textFieldShouldReturn:(UITextField*)theTextField
{
    NSAssert(theTextField == currentField, @"theTextField == currentTextField");
    
    // When the user presses return, take focus away from the text
    // field so that the keyboard is dismissed.
    NSUserDefaults* defaults = [NSUserDefaults standardUserDefaults]; 
    if(theTextField == usernameField)
    {
        [defaults setObject:theTextField.text forKey:usernameKey];
    }
    else if(theTextField == passwordField)
    {
        [defaults setObject:theTextField.text forKey:passwordKey];
    }
    loginButton.enabled = usernameField.text.length > 0;
    [loginButton setAlpha:loginButton.enabled ? 1.0 : 0.5];

    [theTextField resignFirstResponder];
    self.currentField = nil;

    return YES;
}

#pragma mark -

// A touch outside the keyboard dismisses the keyboard, and
// sets back the old field value.
-(void)touchesBegan:(NSSet*)touches withEvent:(UIEvent*)event
{
    [currentField resignFirstResponder];
    currentField.text = oldFieldValue; 
    self.currentField = nil;
    [super touchesBegan:touches withEvent:event];
}

#pragma mark UI Actions

-(IBAction)sslChanged:(id)s
{
    UISwitch* sender = (UISwitch*)s;
    [[NSUserDefaults standardUserDefaults] setBool:sender.isOn forKey:sslKey];
}

#pragma mark Login

-(void)exception:(NSString*)s
{
    [self connecting:FALSE];

    // We always create a new communicator each time
    // we try to login.
    [communicator destroy];
    self.communicator = nil;
 
    loginButton.enabled = usernameField.text.length > 0;
    [loginButton setAlpha:loginButton.enabled ? 1.0 : 0.5];
    
    // open an alert with just an OK button
    UIAlertView *alert = [[UIAlertView alloc] initWithTitle:@"Error"
                                                     message:s
                                                    delegate:nil
                                           cancelButtonTitle:@"OK"
                                           otherButtonTitles:nil];
    [alert show];       
}

-(IBAction)login:(id)sender
{
    ICEInitializationData* initData = [ICEInitializationData initializationData];
    
    initData.properties = [ICEUtil createProperties];
    [initData.properties load:[[[NSBundle mainBundle] resourcePath] stringByAppendingPathComponent:@"config.client"]];
    [initData.properties setProperty:@"Ice.ACM.Client.Timeout" value:@"0"];
    [initData.properties setProperty:@"Ice.RetryIntervals" value:@"-1"];
    
    initData.dispatcher = ^(id<ICEDispatcherCall> call, id<ICEConnection> con)
    {
        dispatch_sync(dispatch_get_main_queue(), ^ { [call run]; });
    };
    
    NSAssert(communicator == nil, @"communicator == nil");
    self.communicator = [ICEUtil createCommunicator:initData];
    
    [self connecting:TRUE];
    
    NSString* username = usernameField.text;
    NSString* password = passwordField.text;

    dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^ {
        @try
        {
            id<GLACIER2RouterPrx> router = [GLACIER2RouterPrx checkedCast:[communicator getDefaultRouter]];
            id<GLACIER2SessionPrx> glacier2session = [router createSession:username password:password];
            id<ChatChatSessionPrx> sess = [ChatChatSessionPrx uncheckedCast:glacier2session];
            
            ICEInt acmTiemout = [router getACMTimeout];
            if(acmTiemout <= 0)
            {
                acmTiemout = (ICEInt)[router getSessionTimeout];
            }
            
            [chatController setup:communicator
                          session:sess
                       acmTimeout:acmTiemout
                           router:router
                         category:[router getCategoryForClient]];
            dispatch_async(dispatch_get_main_queue(), ^ {
                [self connecting:FALSE];
                
                // The communicator is now owned by the ChatController.
                self.communicator = nil;
                
                [chatController activate:@"Chat"];
                [self.navigationController pushViewController:chatController animated:YES];
            });
        }
        @catch(GLACIER2CannotCreateSessionException* ex)
        {
            NSString* s = [NSString stringWithFormat:@"Session creation failed: %@", ex.reason_];
            dispatch_async(dispatch_get_main_queue(), ^ {
                [self exception:s];
            });
        }
        @catch(GLACIER2PermissionDeniedException* ex)
        {
            NSString* s = [NSString stringWithFormat:@"Login failed: %@", ex.reason_];
            dispatch_async(dispatch_get_main_queue(), ^ {
                [self exception:s];
            });
        }
        @catch(ICEEndpointParseException* ex)
        {
            NSString* s = [NSString stringWithFormat:@"Invalid router: %@", ex.reason];
            dispatch_async(dispatch_get_main_queue(), ^ {
                [self exception:s];
            });
        }
        @catch(ICEException* ex)
        {
            dispatch_async(dispatch_get_main_queue(), ^ {
                [self exception:[ex description]];
            });
        }        
    });
}

@end

File: swift/iOS/Chat/LoginController.swift
//
// Copyright (c) ZeroC, Inc. All rights reserved.
//

import Glacier2
import Ice
import PromiseKit
import UIKit

struct Configuration {
    enum Keys: String {
        case user = "usernameKey"
        case password = "passwordKey"
        case ssl = "sslKey"
        case rememberMe = "rememberKey"
    }
}

class LoginController: UIViewController, UITextFieldDelegate, UIAlertViewDelegate {
    @IBOutlet var usernameField: UITextField!
    @IBOutlet var passwordField: UITextField!
    @IBOutlet var loginButton: LoadingButton!
    @IBOutlet var keyboardHeightLayoutConstraint: NSLayoutConstraint?
    @IBOutlet var rememberSwitch: UISwitch!

    weak var currentField: UITextField?
    var communicator: Ice.Communicator?

    var connecting: Bool = false {
        didSet {
            loginButton.loading = connecting
            loginButton.isEnabled = !connecting
            usernameField.isEnabled = !connecting
            passwordField.isEnabled = !connecting
        }
    }

    var remembering: Bool = false {
        didSet {
            if remembering == false { initialize() }
        }
    }

    deinit {
        NotificationCenter.default.removeObserver(self)
    }

    override func viewDidLoad() {
        super.viewDidLoad()
        configureDefaultValues()
        NotificationCenter.default.addObserver(self,
                                               selector: #selector(keyboardWillShow(_:)),
                                               name: UIResponder.keyboardWillShowNotification,
                                               object: nil)
        NotificationCenter.default.addObserver(self,
                                               selector: #selector(keyboardWillHide(_:)),
                                               name: UIResponder.keyboardWillHideNotification,
                                               object: nil)
    }

    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(animated)
        if (usernameField.text ?? "").isEmpty {
            loginButton.isEnabled = false
            loginButton.alpha = 0.5
        } else {
            loginButton.isEnabled = true
            loginButton.alpha = 1.0
        }
    }

    // Observing touchUpInside on UIButton
    @IBAction func login(_: Any) {
        var properties: Properties {
            let prop = Ice.createProperties()
            prop.setProperty(key: "Ice.Plugin.IceSSL", value: "1")
            prop.setProperty(key: "Ice.Default.Router",
                             value: "Glacier2/router:wss -p 443 -h zeroc.com -r /demo-proxy/chat/glacier2 -t 10000")
            prop.setProperty(key: "IceSSL.UsePlatformCAs", value: "1")
            prop.setProperty(key: "IceSSL.CheckCertName", value: "1")
            prop.setProperty(key: "Ice.ACM.Client.Timeout", value: "0")
            prop.setProperty(key: "Ice.RetryIntervals", value: "-1")
            return prop
        }
        var initData = Ice.InitializationData()
        let chatController: ChatController?
        let storyboard = UIStoryboard(name: "Main", bundle: nil)
        chatController = storyboard.instantiateViewController(withIdentifier: "ChatController") as? ChatController
        initData.properties = properties

        connecting = true

        let username: String = usernameField.text ?? ""
        let password: String = passwordField.text ?? ""

        var chatsession: ChatSessionPrx?
        var acmTimeout: Int32 = 0

        do {
            communicator = try Ice.initialize(initData)
            let router = uncheckedCast(prx: communicator!.getDefaultRouter()!, type: Glacier2.RouterPrx.self)

            router.createSessionAsync(userId: username, password: password).then { session -> Promise<Int32> in
                precondition(session != nil)
                chatsession = uncheckedCast(prx: session!, type: ChatSessionPrx.self)
                return router.getACMTimeoutAsync()
            }.then { timeout -> Promise<String> in
                acmTimeout = timeout
                return router.getCategoryForClientAsync()
            }.done { category -> Void in

                try chatController!.setup(communicator: self.communicator!,
                                          session: chatsession!,
                                          acmTimeout: acmTimeout,
                                          router: router,
                                          category: category)

                self.connecting = false

                //
                // The communicator is now owned by the ChatController.
                //

                if self.remembering == true {
                    UserDefaults.standard.set(username, forKey: Configuration.Keys.user.rawValue)
                    UserDefaults.standard.set(password, forKey: Configuration.Keys.password.rawValue)
                }

                chatController?.currentUser = ChatUser(name: username.lowercased())
                self.communicator = nil
                self.navigationController!.pushViewController(chatController!, animated: true)
            }.catch { err in
                if let ex = err as? Glacier2.CannotCreateSessionException {
                    self.exception("Session creation failed: \(ex.reason)")
                } else if let ex = err as? Glacier2.PermissionDeniedException {
                    self.exception("Login failed: \(ex.reason)")
                } else if let ex = err as? Ice.EndpointParseException {
                    self.exception("Invalid router: \(ex)")
                } else {
                    self.exception("Error: \(err)")
                }
            }
        } catch {
            exception("Error: \(error)")
        }
    }

    // Observing touchUpInside on UISwitch
    @IBAction func didTapSwitch(_: UISwitch) {
        remembering = !remembering
        UserDefaults.standard.set(remembering, forKey: Configuration.Keys.rememberMe.rawValue)
    }

    // Observing input changes to UITextFields
    @IBAction func didChange(_: UITextField) {
        validate(username: usernameField.text, password: passwordField.text)
    }

    @objc func dismssKeyboard(_: UITapGestureRecognizer) {
        view.endEditing(true)
    }

    @objc func didEnterBackground() {
        if let communicator = communicator {
            DispatchQueue.global().async {
                communicator.destroy()
            }
        }
    }

    @objc func keyboardWillShow(_ notification: NSNotification) {
        let frame = notification.userInfo?[UIResponder.keyboardFrameBeginUserInfoKey] as? NSValue
        if let keyboardSize = frame?.cgRectValue {
            if view.frame.origin.y == 0 {
                view.frame.origin.y -= keyboardSize.height / 2
            }
        }
    }

    @objc func keyboardWillHide(_: NSNotification) {
        view.frame.origin.y = 0
    }

    func validate(username: String?, password _: String?) {
        if (username ?? "").isEmpty {
            loginButton.isEnabled = false
            loginButton.alpha = 0.5
        } else {
            loginButton.isEnabled = true
            loginButton.alpha = 1.0
        }
    }

    func initialize() {
        UserDefaults.standard.register(defaults: [Configuration.Keys.user.rawValue: "",
                                                  Configuration.Keys.password.rawValue: "",
                                                  Configuration.Keys.ssl.rawValue: true,
                                                  Configuration.Keys.rememberMe.rawValue: false])
    }

    func configureDefaultValues() {
        let bool = UserDefaults.standard.bool(forKey: Configuration.Keys.rememberMe.rawValue)
        remembering = bool
        rememberSwitch.setOn(bool, animated: false)
        if bool == true {
            passwordField.text = UserDefaults.standard.string(forKey: Configuration.Keys.password.rawValue)
            usernameField.text = UserDefaults.standard.string(forKey: Configuration.Keys.user.rawValue)
        }
        let tapGesture = UITapGestureRecognizer(target: self, action: #selector(dismssKeyboard(_:)))
        view.addGestureRecognizer(tapGesture)
        usernameField.clearButtonMode = .whileEditing
        passwordField.clearButtonMode = .whileEditing
    }

    func exception(_ message: String) {
        connecting = false
        if (usernameField.text ?? "").isEmpty {
            loginButton.isEnabled = false
            loginButton.alpha = 0.5
        } else {
            loginButton.isEnabled = true
            loginButton.alpha = 1.0
        }

        let alert = UIAlertController(title: "Error", message: message, preferredStyle: .alert)
        alert.addAction(UIAlertAction(title: "OK", style: .default, handler: nil))
        present(alert, animated: true, completion: nil)
    }
}
